pkg/reco/recommenders.go

```diff
@@ -0,0 +1,173 @@
+)
+
+type RecommendationWorkflow interface {
+	Execute(ctx context.Context, wm WorkloadMeta) (*v1alpha1.HPAConfiguration, *v1alpha1.HPAConfiguration, *Policy, error)
+}
+
+type Recommender interface {
+	Recommend(ctx context.Context, wm WorkloadMeta) (*v1alpha1.HPAConfiguration, error)
+}
+
+type RecommendationWorkflowImpl struct {
+	recommender     Recommender
+	policyIterators map[string]PolicyIterator
+	logger          logr.Logger
+}
+
+type WorkloadMeta struct {
+	metav1.TypeMeta
+	Name      string
+	Namespace string
+}
+
+type RecoWorkflowBuilder RecommendationWorkflowImpl
+
+func (b *RecoWorkflowBuilder) WithRecommender(r Recommender) *RecoWorkflowBuilder {
+	if b.recommender == nil {
+		b.recommender = r
+		return b
+	}
+	golog.Println("Only one recommender must be added. There's already one configured so ignoring this one.")
+	return b
+}
+
+func (b *RecoWorkflowBuilder) WithPolicyIterator(name string, p PolicyIterator) *RecoWorkflowBuilder {
+	if b.policyIterators == nil {
+		b.policyIterators = make(map[string]PolicyIterator)
+		b.policyIterators[name] = p
+	} else if _, ok := b.policyIterators[name]; !ok {
+		b.policyIterators[name] = p
+	}
+	return b
+}
+
+func (b *RecoWorkflowBuilder) WithLogger(log logr.Logger) *RecoWorkflowBuilder {
+	var zeroValLogger logr.Logger
+	if b.logger == zeroValLogger {
+		b.logger = log
+	}
+	return b
+}
+
+func (b *RecoWorkflowBuilder) Build() RecommendationWorkflow {
+	var zeroValLogger logr.Logger
+	if b.logger == zeroValLogger {
+		b.logger = zap.New()
+	}
+	return &RecommendationWorkflowImpl{
+		recommender:     b.recommender,
+		policyIterators: b.policyIterators,
+		logger:          b.logger,
+	}
+}
+
+func NewRecommendationWorkflowBuilder() *RecoWorkflowBuilder {
+	return &RecoWorkflowBuilder{}
+}
+
+type MockRecommender struct {
+	Min       int
+	Threshold int
+	Max       int
+}
+
+func (r *MockRecommender) Recommend(ctx context.Context, wm WorkloadMeta) (*v1alpha1.HPAConfiguration, error) {
+	return &v1alpha1.HPAConfiguration{
+		Min:               r.Min,
+		Max:               r.Max,
+		TargetMetricValue: r.Threshold,
+	}, nil
+}
+
+func (rw *RecommendationWorkflowImpl) Execute(ctx context.Context, wm WorkloadMeta) (*v1alpha1.HPAConfiguration, *v1alpha1.HPAConfiguration, *Policy, error) {
+	ctx = log.IntoContext(ctx, rw.logger)
+	if rw.recommender == nil {
+		return nil, nil, nil, errors.New("No recommenders configured in the workflow.")
+	}
+	targetRecoConfig, err := rw.recommender.Recommend(ctx, wm)
+	if err != nil {
+		rw.logger.Error(err, "Error while generating recommendation")
+		return nil, nil, nil, errors.New("Unable to generate recommendation")
+	}
+	var nextPolicy *Policy
+	for i, pi := range rw.policyIterators {
+		rw.logger.V(0).Info("Running policy iterator", "iterator", i)
+		p, err := pi.NextPolicy(ctx, wm)
+		if err != nil {
+			rw.logger.Error(err, "Error while generating recommendation")
+			return nil, nil, nil, err
+		}
+
+		if p == nil {
```

Comment:
Why are we skipping the noop policy case? if an iterator suggests noop policy, then that should be the one that should get applied.

Reply:
PolicyIterators returning nil is interpreted as that the PI doesn't want to alter the policy configured on the reco or have no opinion on what the policy to be set. Some examples are 

1. Breach analyzer PI not detecting any breach to the workload. (Doesn't want to alter the policy)
2. Default PI not finding any default policies. (No opinion)